src/modules/aggregator/aggregator.c

```diff
@@ -0,0 +1,534 @@
+ */
+static int aggregate_try_sink (flux_t h, struct aggregate *ag)
+{
+    if (aggregate_sink (h, ag) < 0) {
+        flux_watcher_t *w;
+        double t = ag->timeout;
+        if (t <= 1e-3)
+            t = .250;
+        flux_log (h, LOG_INFO, "sink: %s: retry  in %.3fs", ag->key, t);
+        /* On failure, retry just once, then abort */
+        w = flux_timer_watcher_create (flux_get_reactor (h),
+                                       t, 0.,
+                                       aggregate_sink_again,
+                                       (void *) ag);
+        flux_watcher_start (w);
+        return (-1);
+    }
+    zhash_delete (ag->ctx->aggregates, ag->key);
+    return (0);
+}
+
+/*
+ *  Flush aggregate `ag` -- forward entry upstream and destroy it locally.
+ */
+static int aggregate_flush (struct aggregate *ag)
+{
+    flux_t h = ag->ctx->h;
+    int rc;
+    assert (ag->ctx->rank != 0);
+    rc = aggregate_forward (h, ag);
+    zhash_delete (ag->ctx->aggregates, ag->key);
+    return (rc);
+}
+
+static void aggregate_destroy (struct aggregate *ag)
+{
+    struct aggregate_entry *ae = zlist_first (ag->entries);
+    while (ae) {
+        aggregate_entry_destroy (ae);
+        ae = zlist_next (ag->entries);
+    }
+    flux_watcher_destroy (ag->tw);
+    free (ag->key);
+    free (ag);
+}
+
+static void timer_cb (flux_reactor_t *r, flux_watcher_t *tw,
+                      int revents, void *arg)
+{
+    aggregate_flush (arg);
+}
+
+static void aggregate_timer_start (struct aggregator *ctx,
+                                   struct aggregate *ag,
+                                   double timeout)
+{
+    if (ctx->rank != 0) {
+        flux_t h = ctx->h;
+        flux_reactor_t *r = flux_get_reactor (h);
+        ag->tw = flux_timer_watcher_create (r, timeout, 0.,
+                                            timer_cb, (void *) ag);
+        if (ag->tw == NULL)
+            flux_log_error (h, "flux_timer_watcher_create");
+        flux_watcher_start (ag->tw);
+    }
+}
+
+static struct aggregate *
+    aggregate_create (struct aggregator *ctx, const char *key)
+{
+    flux_t h = ctx->h;
+
+    struct aggregate *ag = xzmalloc (sizeof (*ag));
+    memset (ag, 0, sizeof (*ag));
+    ag->ctx = ctx;
+    ag->key = strdup (key);
+    if (!(ag->entries = zlist_new ())) {
+        flux_log_error (h, "zlist_new");
+        aggregate_destroy (ag);
+        return (NULL);
+    }
+    return (ag);
+}
+
+static void aggregator_destroy (struct aggregator *ctx)
+{
+    if (ctx) {
+        if (ctx->aggregates)
+            zhash_destroy (&ctx->aggregates);
+        free (ctx);
+    }
+}
+
+static int attr_get_int (flux_t h, const char *attr)
+{
+    unsigned long l;
+    char *p;
+    const char *s = flux_attr_get (h, attr, 0);
+    if (!s)
+        return (-1);
+    errno = 0;
+    l = strtoul (s, &p, 10);
+    if (*p != '\0' || errno != 0) {
+        flux_log_error (h, "flux_attr_get (%s) = %s", attr, s);
+        return (-1);
```

Comment:
Working through these now, but this one I couldn't follow, sorry.

Are you asking about `attr_get_int` return `-1` on failure? The two callers do seem to check this, but I have to assume I missed your comment here: (-1 indicates error from this function)
```C
static double timer_scale (flux_t h)
{
    long level, maxlevel;
    if (((level = attr_get_int (h, "tbon.level")) < 0) ||
        ((maxlevel = attr_get_int (h, "tbon.maxlevel")) < 0)) {
        return (1.);
    }
    return (maxlevel - level + 1.);
}
```